src/clif-util.rs

```diff
@@ -45,125 +41,179 @@ mod print_cfg;
+        _ => "Invalid command",
+    };
 
-    result
+    SubCommand::with_name(cmd)
+        .about(about_str)
+        .arg(add_verbose_flag())
+        .arg(add_print_flag())
+        .arg(add_time_flag())
+        .arg(add_set_flag())
+        .arg(add_target_flag())
+        .arg(add_input_file_arg())
 }
 
 fn main() {
-    if let Err(mut msg) = clif_util() {
+    let app_cmds = App::new("Cranelift code generator utility")
+        .version(VERSION)
+        .subcommand(
+            SubCommand::with_name("test")
+                .about("Run Cranelift tests")
+                .arg(add_verbose_flag())
+                .arg(add_time_flag())
+                .arg(add_input_file_arg()),
+        )
+        .subcommand(
+            SubCommand::with_name("cat")
+                .about("Outputs .clif file")
+                .arg(add_input_file_arg()),
+        )
+        .subcommand(
+            SubCommand::with_name("print-cfg")
+                .about("Prints out cfg in dot format")
+                .arg(add_input_file_arg()),
+        )
+        .subcommand(
+            add_wasm_or_compile("compile")
+                .arg(
+                    Arg::with_name("just_decode")
+                        .short("t")
+                        .help("Just decode WebAssembly to Cranelift IR"),
+                )
+                .arg(Arg::with_name("check_translation").short("c").help(
+                    "Just checks the correctness of Cranelift IR translated from WebAssembly",
+                )),
+        )
+        .subcommand(add_wasm_or_compile("wasm"));
+
+    let res_util = match app_cmds.get_matches().subcommand() {
+        ("cat", Some(rest_cmd)) => {
+            let mut file_vec: Vec<String> = Vec::new();
+            get_vec(&mut file_vec, rest_cmd.values_of("file"));
+            cat::run(&file_vec)
+        }
+        ("test", Some(rest_cmd)) => {
+            let mut file_vec: Vec<String> = Vec::new();
+            get_vec(&mut file_vec, rest_cmd.values_of("file"));
+            cranelift_filetests::run(rest_cmd.is_present("time-passes"), &file_vec).map(|_time| ())
+        }
+        ("filecheck", Some(rest_cmd)) => {
+            let mut file_vec: Vec<String> = Vec::new();
+            get_vec(&mut file_vec, rest_cmd.values_of("file"));
+            rsfilecheck::run(&file_vec, rest_cmd.is_present("verbose"))
+        }
+        ("print-cfg", Some(rest_cmd)) => {
+            let mut file_vec: Vec<String> = Vec::new();
+            get_vec(&mut file_vec, rest_cmd.values_of("file"));
+            print_cfg::run(&file_vec)
+        }
+        ("compile", Some(rest_cmd)) => {
+            let mut file_vec: Vec<String> = Vec::new();
+            get_vec(&mut file_vec, rest_cmd.values_of("file"));
+
+            let mut set_vec: Vec<String> = Vec::new();
+            get_vec(&mut set_vec, rest_cmd.values_of("set"));
+
+            let mut target_val: &str = "";
+            if let Some(clap_target_vec) = rest_cmd.values_of("target") {
```

Comment:
Here and below, can this use `value_of` rather than `values_of`, so that you don't need the subsequent loop?